src/Storages/NamedCollections.cpp

```diff
@@ -0,0 +1,545 @@
+            copyConfigValue<String>(config_, collection_path + '.' + key, *config, key);
+    }
+
+    template <typename T> T get(const Key & key) const
+    {
+        return getConfigValue<T>(*config, key);
+    }
+
+    template <typename T> T getOrDefault(const Key & key, const T & default_value) const
+    {
+        return getConfigValueOrDefault<T>(*config, key, default_value);
+    }
+
+    template <typename T> void set(const Key & key, const T & value, bool update_if_exists)
+    {
+        setConfigValue<T>(*config, key, value, update_if_exists);
+        if (!keys.contains(key))
+            keys.insert(key);
+    }
+
+    void remove(const Key & key)
+    {
+        removeConfigValue(*config, key);
+        [[maybe_unused]] auto removed = keys.erase(key);
+        assert(removed);
+    }
+
+    Keys getKeys() const
+    {
+        return keys;
+    }
+
+    ImplPtr copy() const
+    {
+        return std::make_unique<Impl>(*this);
+    }
+
+    std::string dumpStructure() const
+    {
+        /// Convert a collection config like
+        /// <collection>
+        ///     <key0>value0</key0>
+        ///     <key1>
+        ///         <key2>value2</key2>
+        ///         <key3>
+        ///            <key4>value3</key4>
+        ///         </key3>
+        ///     </key1>
+        /// </collection>
+        /// to a string:
+        /// "key0: value0
+        ///  key1:
+        ///  	key2: value2
+        ///  	key3:
+        ///  		key4: value3"
+        WriteBufferFromOwnString wb;
+        Strings prev_key_parts;
+        for (const auto & key : keys)
+        {
+            Strings key_parts;
+            splitInto<'.'>(key_parts, key);
+            size_t tab_cnt = 0;
+
+            auto it = key_parts.begin();
+            auto prev_key_parts_it = prev_key_parts.begin();
+            while (it != key_parts.end()
+                   && prev_key_parts_it != prev_key_parts.end()
+                   && *it == *prev_key_parts_it)
+            {
+                ++it;
+                ++prev_key_parts_it;
+                ++tab_cnt;
+            }
+
+            auto start_it = it;
+            for (; it != key_parts.end(); ++it)
+            {
+                if (it != start_it)
+                    wb << '\n';
+                wb << std::string(tab_cnt++, '\t');
+                wb << *it << ':';
+            }
+            wb << '\t' << get<String>(key) << '\n';
+            prev_key_parts = key_parts;
+        }
+        return wb.str();
+    }
+
+private:
+    template <typename T> static T getConfigValue(
+        const Poco::Util::AbstractConfiguration & config,
+        const std::string & path)
+    {
+        return getConfigValueOrDefault<T>(config, path);
+    }
+
+    template <typename T> static T getConfigValueOrDefault(
+        const Poco::Util::AbstractConfiguration & config,
+        const std::string & path,
+        const std::optional<T> & default_value = std::nullopt)
```

Comment:
you construct temp `std::optional` by copying `T` (it can be `std::string`) just to denote optional value.
IMO better way of doing this would be `const T * default_value = nullptr`.

It would be really nice if they managed to make `std::optional<T&>` work 😞